stream_alert/rule_processor/threat_intel.py

```diff
@@ -246,12 +246,22 @@ def _process_ioc(self, ioc_collections):
             ioc_collections (list): A list of StreamIoc instances.
         """
         # Segment data before calling DynamoDB table with batch_get_item.
-        for subset in StreamThreatIntel._segment(ioc_collections):
+        for subset in self._segment(ioc_collections):
             query_values = []
             for ioc in subset:
                 if ioc.value not in query_values:
                     query_values.append(ioc.value)
-            query_result = self._query(query_values)
+
+            query_result = []
+            has_unprocessed_key = True
+            while has_unprocessed_key:
```

Comment:
We probably want to give a limit here so it doesn't run forever 😄 

As an example, we can also optionally stack decorators: https://github.com/airbnb/streamalert/blob/master/stream_alert/rule_processor/handler.py#L273-L301

In your experience, did you ever observe a subset of keys being unprocessed? or is this feature to be proactive?

If anything, you could just feed it back into `query` once, and then backoff on predicate or on exception .  We can chat more tomorrow if that's still unclear